### ISSUE TITLE
windowing support

### ISSUE THREAD

**Issue by manuzhang** [gearpump/gearpump#1034]
We only have `WindowState` now, which limits the usage of windowing to exactly-once processing. A full windowing support should be oblivious to whether it's at-most-once, at-least-once or exactly-once. Also, we need a window function in DSL API. 

Besides, [Google's Dataflow model](http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43864.pdf) introduced the concept of event-time processing, unaligned window / session (while only aligned window is supported now)

> Sessions are windows that capture some period of activity
over a subset of the data, in this case per key. Typically
they are defined by a timeout gap. Any events that occur
within a span of time less than the timeout are grouped
together as a session. Sessions are unaligned windows.

and triggers over window for incremental processing as well as use cases that motivates such concepts

use cases | requirements 
-------------- | -------------------
session-based analysis for search, ads, analytics, social, and YouTube | unaligned windows  
billing pipelines |  triggers and incremental processing support via accumulation and retraction
aggregate statistics calculations | watermark triggers
abuse detection pipelines | percentile watermark triggers
recommendation generation | processing (system) time triggers
anomaly detection | data driven triggers

It's nice for Gearpump to have these augments, and it shouldn't be hard for Gearpump to do so since event-time processing and watermark are already built-in. 

In summary, I'd like to implement the following features for windowing support, 

- [ ] window processor in Graph API for all message processing semantics
- [ ] window function in DSL API for all message processing semantics
- [ ] support unaligned sessions
- [ ] trigger API that includes event-time triggers, watermark triggers and data driven triggers (not sure whether we need processing time triggers)

**Comment by kkasravi** [comment]
Moved to JIRA (including description) - see https://issues.apache.org/jira/browse/GEARPUMP-23
